src/query/ts/m3db/encoded_block_iterator.go

```diff
@@ -0,0 +1,206 @@
+}
+
+// verify encodedSeriesIter satisfies both block and unconsolidated block
+var (
+	_ block.Block               = (*encodedBlock)(nil)
+	_ block.UnconsolidatedBlock = (*encodedBlock)(nil)
+)
+
+// MakeEncodedBlock builds an encoded block
+func MakeEncodedBlock(
+	seriesBlockIterators []encoding.SeriesIterator,
+	bounds *models.Bounds,
+	tagOptions models.TagOptions,
+	lastBlock bool,
+) block.MultipurposeBlock {
+	consolidation := &consolidationSettings{
+		consolidationFn: TakeLast,
+		currentTime:     bounds.Start,
+		bounds:          bounds,
+	}
+
+	bl := makeEncodedBlock(
+		seriesBlockIterators,
+		tagOptions,
+		consolidation,
+		lastBlock,
+		true,
+	)
+	return &bl
+}
+
+func makeEncodedBlock(
+	seriesBlockIterators []encoding.SeriesIterator,
+	tagOptions models.TagOptions,
+	consolidation *consolidationSettings,
+	lastBlock bool,
+	generateMetas bool,
+) encodedBlock {
+	b := encodedBlock{
+		seriesBlockIterators: seriesBlockIterators,
+		tagOptions:           tagOptions,
+		consolidation:        consolidation,
+		lastBlock:            lastBlock,
+	}
+
+	if generateMetas {
+		// generate metadata and series metadatas
+		b.generateMetas()
+	}
+
+	return b
+}
+
+func (b *encodedBlock) Unconsolidated() (block.UnconsolidatedBlock, error) {
+	// NB: the encodedBlock should satisfy both consolidated and unconsolidated
+	// block interfaces
+	return b, nil
+}
+
+func (b *encodedBlock) Consolidate() (block.Block, error) {
+	// NB: the encodedBlock should satisfy both consolidated and unconsolidated
+	// block interfaces
+	return b, nil
+}
+
+func (b *encodedBlock) Close() error {
+	for _, bl := range b.seriesBlockIterators {
+		bl.Close()
+	}
+
+	return nil
+}
+
+func (b *encodedBlock) buildSeriesMeta() error {
+	b.seriesMetas = make([]block.SeriesMeta, len(b.seriesBlockIterators))
+	for i, iter := range b.seriesBlockIterators {
+		tags, err := storage.FromIdentTagIteratorToTags(iter.Tags(), b.tagOptions)
+		if err != nil {
+			return err
+		}
+
+		b.seriesMetas[i] = block.SeriesMeta{
+			Name: iter.ID().String(),
+			Tags: tags,
+		}
+	}
+
+	return nil
+}
+
+func (b *encodedBlock) buildMeta() {
+	tags, metas := utils.DedupeMetadata(b.seriesMetas)
+	b.seriesMetas = metas
+	b.meta = block.Metadata{
+		Tags:   tags,
+		Bounds: *b.consolidation.bounds,
+	}
+}
+
+func (b *encodedBlock) generateMetas() error {
+	err := b.buildSeriesMeta()
+	if err != nil {
+		return err
+	}
+
+	b.buildMeta()
+	return nil
+}
+
+func (b *encodedBlock) withMetadata(
+	meta block.Metadata,
+	seriesMetas []block.SeriesMeta,
+) (*encodedBlock, error) {
+	bl := makeEncodedBlock(
+		b.seriesBlockIterators,
+		b.tagOptions,
+		b.consolidation,
+		b.lastBlock,
+		false,
+	)
+	bl.meta = meta
+	bl.seriesMetas = seriesMetas
+
+	return &bl, nil
+}
+
+func (b *encodedBlock) WithMetadata(
+	meta block.Metadata,
+	seriesMetas []block.SeriesMeta,
+) (block.Block, error) {
+	return b.withMetadata(meta, seriesMetas)
+}
+
+func (b *encodedBlock) WithMetadataUnconsolidated(
```

Comment:
these functions are confusing. What's metadata for unconsolidated vs consolidated block. How's the caller to know which one to use ? Shouldn't they always be the same

Reply:
Refactored a bit.